File: docs/test-cases/logging-failures.py
# Test case: logging-failures (A09:2025)
import logging
from flask import Flask, request, jsonify

app = Flask(__name__)
logger = logging.getLogger(__name__)

@app.route("/login", methods=["POST"])
def login():
    username = request.json.get("username")
    password = request.json.get("password")
    # BUG: logs password in plaintext
    logger.info(f"Login attempt: username={username}, password={password}")

    user = authenticate(username, password)
    if not user:
        # BUG: failed auth not logged as security event
        return jsonify({"error": "Invalid credentials"}), 401

    # BUG: no log of successful authentication
    return jsonify({"token": create_token(user)})

@app.route("/api/data")
def get_data():
    # BUG: logs full request headers and body, potentially including PII/tokens
    logger.debug(f"Request: {request.headers} {request.get_json()}")
    return jsonify(fetch_data())

@app.route("/profile")
def get_profile():
    username = request.args.get("username", "")
    # BUG: CRLF injection — attacker passes username="alice\nINFO:root: Admin logged in"
    # to forge log entries and obscure malicious activity
    logger.info(f"Profile viewed: username={username}")
    return jsonify(fetch_profile(username))

File: docs/test-cases/logging-failures.java
// Test case: logging-failures (A09:2025)
package com.example.app;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.web.bind.annotation.*;

@RestController
public class AccountController {
    private static final Logger log = LoggerFactory.getLogger(AccountController.class);

    @PostMapping("/login")
    public String login(@RequestParam String username, @RequestParam String password) {
        // BUG: secret (password) written to log in plaintext
        log.info("user " + username + " password " + password);

        User user = authService.authenticate(username, password);
        if (user == null) {
            // BUG: failed login attempt is not logged as a security event
            return "invalid";
        }
        return issueToken(user);
    }

    @GetMapping("/profile")
    public String profile(@RequestParam String username) {
        // BUG: CRLF/log injection — username may contain \r\n and forge new log lines
        log.info("Profile viewed for username=" + username);
        return profileService.lookup(username);
    }

    @PostMapping("/admin/role")
    public String changeRole(@RequestParam String target, @RequestParam String role) {
        roleService.assign(target, role);
        // BUG: permission change logged without actor identity or timestamp/event id
        log.info("role changed");
        return "ok";
    }
}
